### PR TITLE
Create a C-style API similar to whisper.cpp

### PR DESCRIPTION
**Pull request by thomasantony**
This change makes it easier to use this code as a library - say to build python bindings on top of it. It extracts out the following functions into `llama.cpp`

- `llama_model_load`
- `llama_eval`
- `llama_model_quantize`

It also moves the relevant struct definitions to `llama.h`. This for example, helps avoid redefinition of `llama_hparams` in `quantize.cpp`. Please let me know if you have any suggestions to improve this.

See [here](https://github.com/thomasantony/llama.cpp/tree/feature/pybind/python) for an example of this library structure in use.

### CODE REVIEW

**Review by ggerganov — CHANGES_REQUESTED** [review]
Yes, this is a step in the right direction, but the exposed things are not the right one.

The `llama_layer` and `llama_model` should not be publicly visible.
You have to wrap them in `llama_context` or `llama_state`, which is just forward declared in `llama.h` file and defined in `llama.cpp`.

See the `whisper.cpp` C-style API for doing it the correct way:
https://github.com/ggerganov/whisper.cpp/blob/master/whisper.h

If you give it another try, make sure to start from latest `master` since things are changing there.

**Comment by thomasantony** [comment]
@ggerganov I have made the changes. Please let me know what you think

File: CMakeLists.txt [hunk 1]
```diff
@@ -78,34 +79,38 @@ endif()
...
 endif()
 
 # if (LLAMA_PERF)
-#     set(LLAMA_EXTRA_FLAGS ${LLAMA_EXTRA_FLAGS} -DGGML_PERF)
+# set(LLAMA_EXTRA_FLAGS ${LLAMA_EXTRA_FLAGS} -DGGML_PERF)
 # endif()
-
 add_executable(llama
     main.cpp
+    llama.cpp
```
Comment (Green-Sky) [inline CMakeLists.txt:113]: llamalib already contains llama.cpp, utils.cpp and utils.h

File: CMakeLists.txt [hunk 2]
```diff
@@ -114,6 +119,9 @@ add_executable(quantize
     utils.cpp
     utils.h)
 
+add_library(llamalib
+    llama.cpp)
```
Comment (Green-Sky) [inline CMakeLists.txt:123]: missing llama.h, utils.cpp and utils.h

**Review by Green-Sky — CHANGES_REQUESTED** [review]

File: llama.h [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    int32_t f16     = 1;
+};
+
+struct llama_context;
+
+void llama_free_context(llama_context* ctx);
+
+const std::vector<gpt_vocab::id>& llama_context_get_embd(const llama_context& ctx);
+gpt_vocab& llama_context_get_vocab(llama_context& ctx);
+bool llama_context_not_finished(const llama_context& ctx);
```
Comment (j-f1) [inline llama.h:46]: Why not `llama_context_not_finished`?

File: llama.h [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    int32_t n_layer = 32;
+    int32_t n_rot   = 64;
+    int32_t f16     = 1;
+};
+
+struct llama_context;
+
+void llama_free_context(llama_context* ctx);
+
+const std::vector<gpt_vocab::id>& llama_context_get_embd(const llama_context& ctx);
```
Comment (j-f1) [inline llama.h:44]: Please expand this unless there’s a strong reason to shorten it:

```suggestion
const std::vector<gpt_vocab::id>& llama_context_get_embedding(const llama_context& ctx);
```

File: llama.h [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+gpt_vocab& llama_context_get_vocab(llama_context& ctx);
+bool llama_context_not_finished(const llama_context& ctx);
+const std::vector<gpt_vocab::id> llama_tokenize_text(const llama_context& ctx, const std::string& text);
+
+const std::vector<gpt_vocab::id>& llama_context_get_last_n_tokens(const llama_context& ctx);
+bool llama_init_context_with_prompt(llama_context& ctx, const std::string& text, bool clear_existing = true);
```
Comment (j-f1) [inline llama.h:50]: Can you organize this header file into categories so it’s easier to read?

1. Things you call once at startup
2. Getters/setters
3. Things that run the main compute loop
4. Teardown code (if any)

File: llama.h [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
+// Various functions for loading a ggml LLaMA model.
+llama_context* llama_init_from_params(const gpt_params& params);
```
Comment (j-f1) [inline llama.h:53]: Why does `llama_init_context_with_prompt` take a `llama_context&` while `llama_init_from_params` returns an `llama_context*`? Can you make these have a similar API, or rename them to clarify how they differ?

File: llama.h [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+// Run inference on a LLaMA model using llama_context.
+std::vector<float> llama_eval(llama_context& ctx, const gpt_params& params, std::string& text);
+
+bool llama_model_quantize(const std::string & fname_inp, const std::string & fname_out, int itype);
+
+bool llama_injest_input(llama_context& ctx, const std::string& text, bool clear_existing = true);
```
Comment (j-f1) [inline llama.h:60]:
```suggestion
bool llama_ingest_input(llama_context& ctx, const std::string& text, bool clear_existing = true);
```

File: llama.h [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+
+bool llama_inference(llama_context& ctx, gpt_vocab::id& model_output);
```
Comment (j-f1) [inline llama.h:62]:
```suggestion
bool llama_infer(llama_context& ctx, gpt_vocab::id& model_output);
```

**Review by j-f1 — COMMENTED** [review]
Some feedback on the API:

File: llama.h (hunk 6 above)
Comment (thomasantony) [reply llama.h]: Removed this confusion in my second pass of refactoring. I feel it is a lot cleaner now. Please take a look.

File: llama.h (hunk 8 above)
Comment (thomasantony) [reply llama.h]: Changed.

File: llama.h (hunk 7 above)
Comment (thomasantony) [reply llama.h]: Fixed this typo here and elsewhere.

File: llama.h (hunk 5 above)
Comment (thomasantony) [reply llama.h]: Done.

File: llama.h (hunk 4 above)
Comment (thomasantony) [reply llama.h]: Updated.

File: llama.h (hunk 3 above)
Comment (thomasantony) [reply llama.h]: Did you mean `llama_context_is_finished`?

File: CMakeLists.txt (hunk 2 above)
Comment (thomasantony) [reply CMakeLists.txt]: Updated CMakelists.

File: CMakeLists.txt (hunk 1 above)
Comment (thomasantony) [reply CMakeLists.txt]: Updated.

**Comment by thomasantony** [comment]
@j-f1 @Green-Sky @ggerganov I have done another pass at refactoring and also fixed a few logical bugs that left interactive mode broken in my original version (among other things). Please take another look. Thanks! I have verified that interactive mode works as intended and inference remains just as fast as before.

File: CMakeLists.txt [hunk 9]
```diff
@@ -80,39 +81,46 @@ endif()
...
     main.cpp
-    utils.cpp
-    utils.h)
+)
 
 add_executable(quantize
     quantize.cpp
+)
+
+add_library(llamalib
```
Comment (Green-Sky) [inline CMakeLists.txt:121]: I feel like we should name the library "llama" and the main executable "main", on linux libraries get prefixed with "lib", so "llamalib" would become "libllamalib", which is a mouth full :sweat_smile: 
@j-f1 @ggerganov ?

Comment (thomasantony) [reply CMakeLists.txt:121]: I would tend to agree. Let see what the chief says. 😄

File: llama.h (hunk 3 above)
Comment (j-f1) [reply llama.h]: That’s what I meant. Sorry for the confusion!

**Comment by ggerganov** [comment]
@thomasantony 
We want to have a C-style API in `llama.h`. We cannot expose C++ constructs

For now, leave it like this and let me apply the necessary changes on top of yours to demonstrate what I have in mind - probably tomorrow or the day after.
Thanks for the contributing!

**Comment by thomasantony** [comment]
> @thomasantony We want to have a C-style API in `llama.h`. We cannot expose C++ constructs
> 
> For now, leave it like this and let me apply the necessary changes on top of yours to demonstrate what I have in mind - probably tomorrow or the day after. Thanks for the contributing!

Okay. Thanks. In the meantime, I will rebase the new changes on the master branch on to this branch.

File: llama.cpp [hunk 10]
```diff
@@ -0,0 +1,1369 @@
...
+    {
+    }
+    ~llama_context(){
+        ggml_free(model.ctx);
+    }
+};
+
+/* Original code by @ggerganov */
+// load the model's weights from a file
+bool llama_model_load(const std::string & fname, llama_model & model, gpt_vocab & vocab, int n_ctx) {
```
Comment (Green-Sky) [inline llama.cpp:131]: I don't see latest changes on [master](https://github.com/ggerganov/llama.cpp/commit/0b366e735729327476ec31da02de3c9c9771ddfb) here, which means you did not update it properly.

Comment (thomasantony) [reply llama.cpp:131]: It's entirely possible I messed something up when merging in changes. The rate at which this repo is updating makes keeping up a Sisyphean task 😅. I will take a look.

Comment (thomasantony) [reply llama.cpp]: I have updated `llama_model_load()`. It was indeed missing a couple of features.

File: CMakeLists.txt (hunk 9 above)
Comment (thomasantony) [reply CMakeLists.txt]: Updated this as suggested.

**Pull request closed without merging**